Added Ctrl-A and Ctrl-M commands

gchp opened the pull request.

gchp committed:
Added Ctrl-A and Ctrl-M commands

kisielk commented on mode/search.go:
```diff
@@ -64,10 +65,7 @@ func (m SearchMode) Draw() {
-	fields := strings.Fields(command)
-	term := fields[0]
-
+func storeSearchTerm(e *editor.Editor, term string) error {
```
This doesn't need to return an error

gchp commented:
Updated - removed the un-needed return value

kisielk commented on mode/normal.go:
```diff
@@ -49,8 +50,14 @@ func (m *normalMode) OnKey(ev *termbox.Event) {
 		case termbox.KeyCtrlA:
-			// TODO: search for next occurrence of word under cursor
-			return
+			// move the cursor past the end of the word.
```
I think we should just fix WordUnderCursor so this is not required. It should work regardless of the position of the cursor within the word.

gchp replied:
I've actually found a few bugs with `WordUnderCursor`. I'm adding some tests to cover them and working on fixes now. Will update the PR when it's ready.

kisielk replied:
Not surprising. I think it's from the original godit code and we don't use it anywhere else.

gchp committed:
Fix WordUnderCursor and add tests

gchp commented:
I've fixed the issues I was seeing with `WordUnderCursor` and it seems to be working as expected now. I added some tests which cover the issues I was seeing.

The one thing I did not fix was the case where the cursor is over a single character word such as `a` or `I`. In this case, the small word, and the next whole word will be returned. I'm not sure how vim handles this, which is why I haven't fixed it. Going to have a look now.

gchp committed:
Make WordUnderCursor work for single character words

gchp commented:
@kisielk I've fixed the single character word case. Let me know if I can improve on this in any way!

kisielk commented:
getting there I think

kisielk commented on buffer/cursor.go:
```diff
@@ -71,6 +71,11 @@ func (c *Cursor) RuneBefore() (rune, int) {
 
+// RuneAfter return the rune after the current cursor and its width in bytes.
+func (c *Cursor) RuneAfter() (rune, int) {
+	return utf8.DecodeRune(c.Line.Data[c.Boffset+1:])
```
what happens when Boffset is on the last character of the line? I think we'll probably get a panic.

There should also be separate tests for this method as well.

kisielk commented on buffer/cursor.go:
```diff
@@ -246,15 +251,29 @@ func (c *Cursor) MoveEOL() {
 	end, beg := *c, *c
 	r, rlen := beg.RuneBefore()
-	if r == utf8.RuneError {
+	if r == utf8.RuneError && !beg.BOL() {
```
I think we just need the `beg.BOL()` check once at the beginning and we can skip all the `RuneBefore` logic until L263 in that case. eg:

```
if !beg.BOL() {
    r, rlen := beg.RuneBefore()
    //...
}

end.Boffset = beg.Boffset
```

kisielk commented on buffer/cursor_test.go:
```diff
@@ -300,3 +300,45 @@ func TestSortCursors(t *testing.T) {
+		t.Error("Incorrect word:", word)
+	}
+
+	// check that if the cursor is at the start of the line
```
should also be a test for the last character of the line

kisielk commented:
what happens when the cursor is on whitespace?

gchp committed:
Generally improved WordUnderCursor & fixed panic with whitespace

gchp committed:
Added tests for Cursor.RuneAfter & fixed end of line error

gchp committed:
Fixed WordUnderCursor panic when on last word in a line

gchp commented:
Applied the changes you suggested.

Currently, when you are on whitespace it will just return `nil`. I think that is reasonable?

kisielk commented on buffer/cursor_test.go:
```diff
@@ -300,3 +301,87 @@ func TestSortCursors(t *testing.T) {
 	}
 }
+
+func TestRuneAfter(t *testing.T) {
```
Please also test the byte length returned

kisielk commented:
I think returning nil is fine for now. Not sure what other vi's do yet.

kisielk commented:
But you probably need to change the implementation of keybindings to check for `nil` so they don't clobber the search term and execute the search command.

gchp committed:
Test byte width returned from RuneAfter is correct

gchp commented:
Added a test for the byte width. What bytes would be longer than 1? I've not come across any.

I don't think that will be an issue. The expected return type of `WordUnderCursor` is `[]byte`. If you return nil, it essentially returns nothing.

I also then convert this to a string in order to pass it to `storeSearchTerm`. As far as I know converting `nil` to a string doesn't yield anything.

See here for an example: http://play.golang.org/p/obFGhvqoQn

kisielk commented:
Any character outside of the latin character set will have a width greater than one, but I don't think we need to explicitly test that, just that we return 0 when we expect it and 1 on a regular character.

As for the nil, I know it works and I'm not concerned with crashes, but more the fact that I think doing a "search for word under cursor" when your cursor is not over a word should not cause the editor to forget the previous search term.

gchp committed:
Don't clobber the last search term if attempting to search for whitespace

gchp commented:
Ah, right yeah that makes sense. Just updated.
